### PR TITLE
4.0.0

### PR DESCRIPTION
**Pull request by rdev**
This is a WIP for the 4.0.0 (Node.js version of the library). It removes browser support completely and makes `now-client` only work in Node

Todo:
- [x] Single file/file group deployments
- [x] Update readme
- [ ] Tests

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,50 +3,76 @@
...
-    "eslint": "^5.16.0",
-    "npm-run-all": "^4.1.5",
-    "rollup": "^1.11.3",
-    "rollup-plugin-babel": "^4.3.2",
-    "rollup-plugin-babel-minify": "^8.0.0",
-    "rollup-plugin-commonjs": "^9.3.4",
-    "rollup-plugin-filesize": "^6.0.1",
-    "rollup-plugin-node-resolve": "^4.2.3",
-    "serve": "^11.0.0"
+    "@types/async-retry": "^1.4.1",
```
Comment (leo) [inline package.json]: Please lock all the dependencies by removing the `^` and adding [this file](https://github.com/zeit/serve/blob/master/.yarnrc).

Comment (leo) [reply package.json]: Both devDependencies and regular ones.

File: src/deployment.ts [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+import pkg from '../package.json'
+
+const getDefaultName = (files: Map<string, DeploymentFile>): string => {
+  const filePath = Array.from(files.values())[0].names[0]
+  const segments = filePath.split('/')
+
+  return segments[segments.length - 1]
+}
+
+export default class Deployment extends EventEmitter {
```
Comment (leo) [inline src/deployment.ts]: Why did you build a class again? Because of the Now CLI code? If so, that's fine.

File: src/deployment.ts [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    return { deployment: json }
+  }
+
+  checkDeploymentStatus = async (): Promise<void> => {
+    if (!this._data) {
+      return
+    }
+
+    // Get builds and deployment status
+    const [deploymentData, buildsData] = await Promise.all([
```
Comment (leo) [inline src/deployment.ts]: Technically, the correct way would be only starting to poll the deployment doc once the builds have finished (like Now CLI does it). This way, you don't waste computational resources.

Comment (leo) [reply src/deployment.ts]: [Here we check](https://github.com/zeit/now-cli/blob/canary/src/commands/deploy/latest.js#L561)
[Here we print](https://github.com/zeit/now-cli/blob/canary/src/commands/deploy/latest.js#L609)

File: src/utils/index.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { DeploymentFile } from "./hashes"
+import { readFile } from "fs-extra"
+import { join } from "path"
+
+export const API_FILES = 'https://api.zeit.co/v2/now/files'
+export const API_DEPLOYMENTS = 'https://api.zeit.co/v8/now/deployments'
```
Comment (leo) [inline src/utils/index.ts]: This needs to be v9 (see [API docs](https://zeit.co/docs/api#endpoints/deployments/create-a-new-deployment)).

File: types/types.d.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  status?: number;
+  methods?: string[];
+}
+
+declare interface Build {
+  src: string;
+  use: string;
+}
+
+declare interface ZEITDeployment {
```
Comment (leo) [inline types/types.d.ts]: Why not `Deployment`?

File: src/deployment.ts (hunk 2 above)
Comment (rdev) [reply src/deployment.ts]: Yep, that's how CLI does it and we need to extend `EventEmmitter`
The usage of the library is still functional

File: types/types.d.ts (hunk 5 above)
Comment (rdev) [reply types/types.d.ts]: When used in type definitions, it conflicts with the `Deployment` class name

File: src/deployment.ts [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+      }
+
+      this.builds[build.id] = build
+    })
+
+    let allBuildsCompleted = true
+    Object.keys(this.builds).forEach((key: string): void => {
+      const build = this.builds[key]
+
+      if (build.readyState !== 'READY') {
```
Comment (leo) [inline src/deployment.ts]: `READY` is not the [only completed state](https://github.com/zeit/now-cli/blob/canary/src/commands/deploy/latest.js#L587)

File: src/deployment.ts [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+      const deploymentUpdate = await deploymentData.json()
+  
+      // Fire deployment state change listeners if needed
+      if (deploymentUpdate.readyState !== this._data.readyState) {
+        this.emit('deployment-state-changed', deploymentUpdate)
+      }
+  
+      this._data = deploymentUpdate
+  
+      if ((this._data as ZEITDeployment).readyState === 'READY') {
```
Comment (leo) [inline src/deployment.ts]: Same for deployment state, use the functions from CLI

**Pull request merged**